Scripting: Cache script results if deterministic

stu-elastic opened the pull request:
Cache results from queries that use scripts if they use only
deterministic API calls.  Nondeterministic API calls are marked in the
whitelist with the `@nondeterministic` annotation.  Examples are
`Math.random()` and `new Date()`.

Refs: #49466

stu-elastic committed:
Notes

stu-elastic committed:
Wiring from whitelist annotations to PainlessMethod and PainlessConstructor

stu-elastic committed:
Generate isResultDeterministic

stu-elastic committed:
formatting, parseInt is deterministic

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into feature/painless-caching-part-2

stu-elastic committed:
WIP

stu-elastic committed:
update whitelist based on core-infra discussion, map/set iterators are ok

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into feature/painless-caching-part-2

stu-elastic committed:
Revert LinkedHashMap

stu-elastic committed:
MaxAggregatorTests.testDontCacheScripts -> testCacheScripts

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Scripting)

stu-elastic committed:
remove stale TODOs

stu-elastic committed:
handle imported and static methods

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into feature/painless-caching-part-2

jdconrad commented on modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/annotation/NonDeterministicAnnotation.java:
```diff
@@ -0,0 +1,29 @@
+
+package org.elasticsearch.painless.spi.annotation;
+
+public class NonDeterministicAnnotation {
```
@rjernst Given our new approach for determining cacheability, I'm not sure this deterministic is a good name for this annotation anymore. Any thoughts?

rjernst replied:
I think this name is good. Using "nocache" or somethign like that here would to me imply the method results were cacheable, while the method being non deterministic implies the script results are not cacheable.

jdconrad commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/java.util.txt:
```diff
@@ -574,8 +574,8 @@ class java.util.Collections {
   void rotate(List,int)
-  void shuffle(List)
-  void shuffle(List,Random)
+  void shuffle(List) @nondeterministic
```
This could run into the shuffle is called on a def problem for runtime resolution where we could possibly miss the randomized result. Though, given how loose we're being with what's cacheable, maybe we should ignore this specific instance for now.

jdconrad replied:
Oops! Never mind as these are static methods. Misread that as List would inherit these.

jdconrad reviewed (approved):
LGTM once we address the couple of comments I left that require further discussion.

rjernst commented on modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/annotation/NonDeterministicAnnotationParser.java:
```diff
@@ -0,0 +1,40 @@
+    private NonDeterministicAnnotationParser() {}
+
+    @Override
+    public Object parse(Map<String, String> arguments) {
```
Unrelated to this PR, but we should consider moving the parser instance into the annotation class itself. Having 2 tiny classes makes navigation when debugging code more difficult IMO.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/ScriptRoot.java:
```diff
@@ -72,4 +73,6 @@ public FunctionTable getFunctionTable() {
         return prefix + "$synthetic$" + syntheticCounter++;
     }
+
+    public void setDeterministic(boolean deterministic) { this.deterministic &= deterministic; }
```
This method name is deceptive since it doesn't really set the deterministic flag. I wonder if this should be something like `markNonDeterministic()`, and does the set to false internally when that is called.

stu-elastic replied:
Very nice. Done.

rjernst reviewed (approved):
LGTM. Left a few suggestions.

stu-elastic commented:
Currently debugging NPE when using scripts for weighted avg, fieldName coming in as null.

stu-elastic committed:
setDeterministic -> markNonDeterministic

stu-elastic commented:
Ok, now that scripts are cachable, here's what's causing the NPE

[`SearchService.loadOrExecuteQueryPhase`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/SearchService.java#L336) 

calls into 

[`IndicesService.loadIntoContext`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/indices/IndicesService.java#L1298) (instead of queryPhase.execute) 

which calls 

[`ShardSearchRequest.cacheKey`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java#L283)

 which eventually calls 

[`MultiValuesSourceFieldConfig.writeTo`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceFieldConfig.java#L107)

 which nulls out because there is no `fieldName`.

Here's the series of calls to reproduce:

```
curl -H 'Content-Type: application/json' -XPUT --user 'elastic-admin:elastic-password' localhost:9200/exams -d '{ "mappings": { "properties": { "grade": { "type": "byte" } } } }'

curl -H 'Content-Type: application/json' -XPOST --user 'elastic-admin:elastic-password' localhost:9200/exams/_doc -d '{"grade": 100, "weight": 2}'

curl -H 'Content-Type: application/json' -XPOST --user 'elastic-admin:elastic-password' localhost:9200/exams/_doc -d '{"grade": 50, "weight": 3}'

curl -H 'Content-Type: application/json' -XPOST --user 'elastic-admin:elastic-password' localhost:9200/exams/_search -d '{
    "size": 0,
    "aggs" : {
        "weighted_grade": {
            "weighted_avg": {
                "value": {
                    "script": "doc.grade.value + 1"
                },
                "weight": {
                    "script": "doc.weight.value + 1"
                }
            }
        }
    }
}'
```

stu-elastic commented:
This just happened to be caught by a test.  Are there other places where `cacheKey` calls `writeTo` with fields that are not available if a script is used?

stu-elastic committed:
MultiValuesSourceFieldConfig, use {read,write}OptionalString for fieldName

stu-elastic commented:
@jpountz the Null Pointer Exception when creating the cache key, mentioned [above](https://github.com/elastic/elasticsearch/pull/50106#issuecomment-565229366) was a bit concerning.  Can you take a look?

stu-elastic commented:
@elasticmachine run elasticsearch-ci/1 elasticsearch-ci/2

stu-elastic commented:
@elasticmachine run elasticsearch-ci/2

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into feature/painless-caching-part-2

stu-elastic committed:
this.deterministic &= deterministic -> this.deterministic &= !nondeterministic;

stu-elastic committed:
guard writeOptional against pre-v7.6.0

jpountz commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/java.lang.txt:
```diff
@@ -729,7 +729,7 @@ class java.lang.StrictMath {
+  double random() @nondeterministic
   double rint(double)
   long round(double)
   double scalb(double,int)
```
what about `System#currentTimeMillis` and `System#nanoTime` that also occur in that file. Should we annotate them too?

stu-elastic replied:
Absolutely, I'm taking another pass at these files now in case there are other misses.

jpountz commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/java.util.txt:
```diff
@@ -574,8 +574,8 @@ class java.util.Collections {
-  void shuffle(List)
-  void shuffle(List,Random)
+  void shuffle(List) @nondeterministic
+  void shuffle(List,Random) @nondeterministic
```
maybe that one is ok since it takes an explicit random?

stu-elastic replied:
Agreed. Reverted.

jpountz commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/java.util.txt:
```diff
@@ -910,22 +910,22 @@ class java.util.PriorityQueue {
-  int nextInt(int)
-  long nextLong()
+  () @nondeterministic
+  (long) @nondeterministic
```
maybe the unseeded constructor is the only one that needs to be annotated?

stu-elastic replied:
Yup.

stu-elastic committed:
disable bwc testing until 7.6 backport

jpountz commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/org.elasticsearch.score.txt:
```diff
@@ -25,8 +25,8 @@ class org.elasticsearch.script.ScoreScript @no_import {
-    double randomScore(org.elasticsearch.script.ScoreScript, int, String) bound_to org.elasticsearch.script.ScoreScriptUtils$RandomScoreField
-    double randomScore(org.elasticsearch.script.ScoreScript, int) bound_to org.elasticsearch.script.ScoreScriptUtils$RandomScoreDoc
+    double randomScore(org.elasticsearch.script.ScoreScript, int, String) bound_to org.elasticsearch.script.ScoreScriptUtils$RandomScoreField @nondeterministic
+    double randomScore(org.elasticsearch.script.ScoreScript, int) bound_to org.elasticsearch.script.ScoreScriptUtils$RandomScoreDoc @nondeterministic
```
These ones are deterministic, the behavior is the same is you always provide the same seed?

stu-elastic replied:
:+1:

stu-elastic committed:
add pr link

stu-elastic committed:
System.nanoTime/currentTimeMillis non, Collections.shuffle with rand det, new Random with seed det, remove scorescriptutils non

stu-elastic committed:
Update script cache testing in HDRPercentilesIT, TDigestPercentileRanksIT andCardinalityIT

stu-elastic committed:
Add MockDeterministicScript and use it in MockScriptEngine.

Add nonDeterministicPluginScripts in MockScriptPlugin as default,
overrideable
method.  If a tests wants to ensure non-deterministic scripts are
cached, then
it can return a non-empty collection here and use the script later.

Allows tests to add deterministic->cacheable tests for scripts.

Add script cache testings to DateHistorgramIT.

Revert AvgAggregatorTests, CardinalityIT, ExtendedStatsIT,
HDRPercentilesIT,
MaxAggregatorTests, ScriptedMetricIT, TDigestPercentileRanksIT

to prepare to reimplement

stu-elastic committed:
Update DateRangeIT

stu-elastic committed:
Update DoubleTermsIT

stu-elastic committed:
Update HistogramIT

stu-elastic committed:
Update LongTermsIT

stu-elastic committed:
Update RangeIT

stu-elastic committed:
Update SignificantTermsSignificanceScoreIT

stu-elastic committed:
Update StringTermsIT

stu-elastic committed:
Update AvgAggregatorTests

stu-elastic committed:
CardinalityIT

stu-elastic committed:
Update ExtendedStatsIT

stu-elastic committed:
Update HDRPercentileRanksIT

stu-elastic committed:
Update HDRPercentilesIT

stu-elastic committed:
Update MaxAggregatorTests MedianAbsoluteDeviationIT

stu-elastic committed:
Update MinIT

stu-elastic committed:
Update ScriptedMetricIT

stu-elastic committed:
Update StatsIT

stu-elastic committed:
Update SumIT

stu-elastic committed:
Update TDigestPercentileRanksIT

stu-elastic committed:
Update TDigestPercentilesIT

stu-elastic committed:
Update TopHitsIT

stu-elastic committed:
Update ValueCountIT

stu-elastic committed:
MinIT merged into MinAggregatorTests

stu-elastic committed:
Update MinAggregatorTests

stu-elastic committed:
Use reproducable random methods instead of Math.random

stu-elastic committed:
Remove unnecessary publics, rearrange imports, correct indent

stu-elastic committed:
Mark Calendar and Clock methods nondeterministic

stu-elastic commented:
@elasticmachine run elasticsearch-ci/1

jdconrad commented:
@stu-elastic Thanks again for all the work getting this done. The tests look especially good.
